LDR Tutorial & Manage LDR pages

kathancox opened the pull request:
## Preview: 
[Setup Guide](https://deploy-preview-19043--cockroachdb-docs.netlify.app/docs/v24.3/set-up-logical-data-replication.html)
[Manage LDR Ref](https://deploy-preview-19043--cockroachdb-docs.netlify.app/docs/v24.3/manage-logical-data-replication.html)

This PR is 1 of 4(?) for LDR docs. It contains:

- Setup Guide (tutorial) for setting up a unidirectional/bidirectional LDR stream. 
- Manage LDR for understanding conflict resolution, working with the DLQ, coordinating schema changes, and job management. Also, a small section on schema validation (this should maybe move into a known limitation section when there is more content).

Fixes DOC-10483, DOC-11073, DOC-10859

Reviewers: You'll see a lot of comments for follow-up PRs to connect pages with links and such.

**Content pages to come:**
LDR Overview
Monitoring
`CREATE`/`SHOW` sql ref
Technical Overview
LDR Metrics Dashboard docs

alicia-l2 commented:
This lgtm from an structure perspective!

alicia-l2 commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,189 @@
+
+In this tutorial, you will set up **logical data replication (LDR)** between two CockroachDB clusters. Both clusters are active and can serve traffic. You can apply the outlined steps to create a _unidirectional_ LDR stream from a source to a destination (cluster A to cluster B). 
+
+To create a _bidirectional_ LDR stream, you can complete the [optional step](#step-4-optional-set-up-bidirectional-ldr-streams) to start a second LDR stream that sends writes from the table on cluster B to the table on cluster A. In a bidirectional setup, one cluster operates as both a source and a destination in separate LDR jobs. 
```
can we be consistent and either say "LDR jobs" or "LDR streams"? Personally I prefer stream. What do you think?

kathancox replied:
Yes, I can take a look at this and be more strict. I think defaulting to `stream` is fine, but there will be some places with `job` when we're talking about viewing jobs etc. It also tracks with the syntax we have in place `create logical replication stream` + `show logical replication jobs`. 

I'll go through and clean up the different situations for these though.

kathancox replied:
Either gone through and adjusted mostly to plain LDR. I've used "stream" as verb for the majority, though I think there are some instances where saying LDR stream just reads easier. However, since we need to talk about looking at and monitoring jobs, I have kept LDR job in quite a few places. Happy to keep adjusting this as we work through the doc set for LDR prior to release.

alicia-l2 commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,189 @@
+
+## Tutorial overview
+
+If you're setting up a bidirectional LDR stream, both clusters will act as a source and a destination in the respective LDR streams. The high-level steps are:
```
"If you're planning on creating a bidirectional setup with LDR?" Bidirectional replication?

kathancox replied:
I've adjusted to "If you're setting up bidirectional LDR"... I think it would be great to just have "uni/bidirectional LDR" as the description of those setups, so we don't have to describe it in every sentence. I think we can define this more in the technical overview. Lmk if you want to rethink this still.

alicia-l2 commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,189 @@
+{% include feature-phases/preview.md %}
+{{site.data.alerts.end}}
+
+In this tutorial, you will set up **logical data replication (LDR)** between two CockroachDB clusters. Both clusters are active and can serve traffic. You can apply the outlined steps to create a _unidirectional_ LDR stream from a source to a destination (cluster A to cluster B). 
```
I think we should call out here somewhere that you will be replicating a table

kathancox replied:
I have adjusted this whole first paragraph. PTAL

alicia-l2 commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,189 @@
+    {{site.data.alerts.end}}
+
+- All nodes in each cluster will need access to the Certificate Authority for the other cluster. Refer to [Connect from the destination to the source](#step-2-connect-from-the-destination-to-the-source).
+- If both clusters are empty, create a table with identical schema definitions (excluding indexes) on both clusters. If one cluster already has an existing table, ensure the other cluster's table definition matches. For more details on the supported schemas, refer to [Schema Validation]({% link {{ page.version.version }}/manage-logical-data-replication.md %}#schema-validation).
```
Can we call out here that you're creating the table that you want to replicate?

alicia-l2 commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,189 @@
+
+_Modes_ determine how LDR replicates the data to the destination cluster. There are two modes:
+
+- `immediate` (default): Inserts data at the [KV layer]({% link {{ page.version.version }}/architecture/storage-layer.md %}), which means that the stream does not encounter issues with updates and deletes. `immediate` mode does not insert at the [SQL layer]({% link {{ page.version.version }}/architecture/sql-layer.md %}), which means that the replication will not follow SQL constraints. As a resu ...
```
@msbutler do you have a suggestion on how this might be better stated? "stream does not encounter issues with updates and deletes." I dont want to give the vibe that the stream just never works with updates and deletes, but we have that weird update/delete conflict in SQL mode

kathancox replied:
It'd be great to get these descriptions in perfect shape so we can add to an include file — we'll need these descriptions on other LDR pages.

alicia-l2 commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,189 @@
+_Modes_ determine how LDR replicates the data to the destination cluster. There are two modes:
+
+- `immediate` (default): Inserts data at the [KV layer]({% link {{ page.version.version }}/architecture/storage-layer.md %}), which means that the stream does not encounter issues with updates and deletes. `immediate` mode does not insert at the [SQL layer]({% link {{ page.version.version }}/architecture/sql-layer.md %}), which means that the replication will not follow SQL constraints. As a resu ...
+- `validated`: Inserts data at the SQL layer. All SQL constraints are adhered to. If a row is unable to apply due to a SQL constraint, it will be put in the [dead letter queue (DLQ)]({% link {{ page.version.version }}/manage-logical-data-replication.md %}#dead-letter-queue-dlq) for manual resolution. SQL does not recognize deletion records used by [MVCC]({% link {{ page.version.version }}/archite ...
```
nit: should this be one sentence "inserts data at the SQL layer, so all SQL constraints are adhered to."

kathancox replied:
Done!

alicia-l2 commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,189 @@
+
+## Step 4. (Optional) Set up bidirectional LDR streams
+
+At this point, you've set up one LDR stream from cluster A as the source to cluster B as the destination. 
```
More general comment, do we want to start talking about all the clusters as in Cluster A and Cluster B, or stick with source/dst?

kathancox replied:
I think because we're doing bidirectional, I think it makes sense to keep both. It is hard to describe using just A and B, because that is pretty abstracted away from the direction flow. However, if you then try to just describe in terms of source/destination, you'll run into situations where you have to say "From the original destination cluster", which gets pretty ugly, very quickly.

alicia-l2 commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,189 @@
+
+    {% include_cached copy-clipboard.html %}
+    ~~~ sql
+    SELECT COUNT(*) from {table_name};
```
did you add this primarily so the user can see if rows are replicating?

kathancox replied:
OK, for now, took that out. Will build out this step with the verification tooling etc, when I'm working on the PR for monitoring LDR.

alicia-l2 commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,189 @@
+1. Access the DB Console at `http://{node IP or hostname}:8080` and enter your user's credentials.
+1. Navigate to the [**Jobs** page]({% link {{ page.version.version }}/ui-jobs-page.md %}) to view a list of all jobs. Use the job **Type** dropdown and select **Replication Producer**. On the source cluster for an LDR stream, this will display the history retention job. This will run while the LDR stream is active to protect changes to the table from [garbage collection]({% link {{ page.version.v ...
+1. Use the job **Type** dropdown and select **Logical Replication Ingestion**. On the destination cluster for an LDR stream, this page will display the logical replication stream job. There will be a progress bar in the **Status** column when the LDR stream is replicating a table with existing data. This progress bar shows the status of the initial scan, which backfills the destination table with ...
+1. Click on **Metrics** in the left-hand navigation menu. Use the **Dashboard** dropdown to select **Logical Data Replication**. This page shows graphs for monitoring the LDR stream. 
```
Do you plan on also adding the list of metrics that you can see on DB console?

kathancox replied:
See the comment below this line. Plan to add the details to the monitoring page. But can add a shortened form here too.

kathancox replied:
I'll do this with the monitoring PR, I think. So, then I know what I'm talking about better. Happy to take a list though if you want to add that here in this PR.

kathancox commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,189 @@
+    (1 row)
+    ~~~
+
+    If you're setting up bidirectional LDR streams, both clusters will have a history retention job and an LDR job running.
```
I'm going to describe this in the technical overview and overview + then link out from. I don't know if we should add too much descriptions here.

kathancox commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,189 @@
+
+## Step 3. Start the LDR stream
+
+In this step, you'll start the LDR stream. You can replicate one or multiple tables in a single LDR stream. You cannot replicate system tables in LDR streams, which means that you must manually apply configurations and cluster settings, such as [row-level TTL]({% link {{ page.version.version }}/row-level-ttl.md %}) and user permissions on the destination cluster.
```
Yes, I have added the destination cluster in here. Also, I went through every step and at each substep where the cluster changes I have bolded source or destination, or both.

kathancox commented:
@alicia-l2 I worked through your feedback. PTAL. I have also added the known limitations, and added more detail in places where the support was unlikely to be a known limitation.
@msbutler Would appreciate your review on these two starting pages.
Thanks both!

alicia-l2 commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -87,10 +87,9 @@ You'll need:
 In this step, you'll set up an external connection from the destination cluster to the source cluster. Depending on how you manage certificates, you must ensure that all nodes between the clusters have access to the certificate of the other cluster.
 
 You can use the `cockroach encode-uri` command to generate a connection string containing a cluster's certificate.
+LDR from the destination cluster.
```
This appeared like a random sentence/not finished

kathancox replied:
Oh weird, will take a look. Thanks for finding!

alicia-l2 commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -103,27 +102,27 @@ For example, in this tutorial, you'll need a connection string for the source cl
-- `immediate` (default): Inserts data at the [KV layer]({% link {{ page.version.version }}/architecture/storage-layer.md %}), which means that the stream does not encounter issues with updates and deletes. `immediate` mode does not insert at the [SQL layer]({% link {{ page.version.version }}/architecture/sql-layer.md %}), which means that the replication will not follow SQL constraints. As a resu ...
-- `validated`: Inserts data at the SQL layer. All SQL constraints are adhered to. If a row is unable to apply due to a SQL constraint, it will be put in the [dead letter queue (DLQ)]({% link {{ page.version.version }}/manage-logical-data-replication.md %}#dead-letter-queue-dlq) for manual resolution. SQL does not recognize deletion records used by [MVCC]({% link {{ page.version.version }}/archite ...
+- `immediate` (default): Inserts data at the [KV layer]({% link {{ page.version.version }}/architecture/storage-layer.md %}), which means that LDR does not encounter issues with updates and deletes. `immediate` mode does not insert at the [SQL layer]({% link {{ page.version.version }}/architecture/sql-layer.md %}), which means that the replication will not follow SQL constraints. As a result, you ...
+- `validated`: Inserts data at the SQL layer, which means that all SQL constraints are adhered to. If a row is unable to apply due to a SQL constraint, it will be put in the [dead letter queue (DLQ)]({% link {{ page.version.version }}/manage-logical-data-replication.md %}#dead-letter-queue-dlq) for manual resolution. SQL does not recognize deletion records used by [MVCC]({% link {{ page.version.v ...
```
Wow you described this update/delete constraint so much better than i did -- thanks!!!

alicia-l2 commented on src/current/v24.3/manage-logical-data-replication.md:
```diff
@@ -0,0 +1,149 @@
+
+Once you have **logical data replication (LDR)** running, you will need to track and manage certain parts of the job:
+
+- [Conflict resolution](#conflict-resolution): As changes to a table replicate from the source to the destination cluster, there can be conflicts during some operations that the job will handle with conflict resolution. When LDR is started, the job creates a [_dead letter queue (DLQ)_](#dead-letter-queue-dlq) table with each replicating table. LDR will send any unresolved conflicts to the DLQ, wh ...
```
could we separate DLQ from conflict resolution? the DLQ also manages "conflicts" but its actually a different definition of "conflict," like a secondary index "conflict"

alicia-l2 replied:
Actually nvm. I see how you talked about it later.

alicia-l2 commented on src/current/v24.3/manage-logical-data-replication.md:
```diff
@@ -0,0 +1,149 @@
+Once you have **logical data replication (LDR)** running, you will need to track and manage certain parts of the job:
+
+- [Conflict resolution](#conflict-resolution): As changes to a table replicate from the source to the destination cluster, there can be conflicts during some operations that the job will handle with conflict resolution. When LDR is started, the job creates a [_dead letter queue (DLQ)_](#dead-letter-queue-dlq) table with each replicating table. LDR will send any unresolved conflicts to the DLQ, wh ...
+- [Schema changes](#schema-changes): The tables that are part of the LDR job may require schema changes, which need to be coordinated. There are some schema changes that are supported while LDR is active.
```
which need to be manually coordinated

kathancox replied:
Added

alicia-l2 commented on src/current/v24.3/manage-logical-data-replication.md:
```diff
@@ -0,0 +1,149 @@
+
+### Dead letter queue (DLQ)
+
+When the LDR job starts, it will create a DLQ table with each replicating table so that unresolved conflicts can be tracked. The DLQ will contain the writes that LDR cannot apply after the retry period. For example, [foreign key]({% link {{ page.version.version }}/foreign-key.md %}) dependendies that are not met when there are foreign key constraints in the schema.
```
small typo in dependencies

kathancox replied:
thanks, corrected

alicia-l2 commented on src/current/v24.3/manage-logical-data-replication.md:
```diff
@@ -0,0 +1,149 @@
+
+### Dead letter queue (DLQ)
+
+When the LDR job starts, it will create a DLQ table with each replicating table so that unresolved conflicts can be tracked. The DLQ will contain the writes that LDR cannot apply after the retry period. For example, [foreign key]({% link {{ page.version.version }}/foreign-key.md %}) dependendies that are not met when there are foreign key constraints in the schema.
```
Can we call out the DLQ table is given a random number to dedupe/ensure that it is unique? users might ask

kathancox replied:
Yes, done this after the `show tables` command where the output shows the DLQ naming

alicia-l2 commented on src/current/v24.3/manage-logical-data-replication.md:
```diff
@@ -0,0 +1,149 @@
+
+### Dead letter queue (DLQ)
+
+When the LDR job starts, it will create a DLQ table with each replicating table so that unresolved conflicts can be tracked. The DLQ will contain the writes that LDR cannot apply after the retry period. For example, [foreign key]({% link {{ page.version.version }}/foreign-key.md %}) dependendies that are not met when there are foreign key constraints in the schema.
```
"For example, [foreign key](https://deploy-preview-19043--cockroachdb-docs.netlify.app/docs/v24.3/foreign-key) dependendies that are not met when there are foreign key constraints in the schema." 

can we add "in validated mode" because this is only a thing in validated mode?

kathancox replied:
Added "in `validated` mode"

alicia-l2 commented on src/current/v24.3/manage-logical-data-replication.md:
```diff
@@ -0,0 +1,152 @@
+
+    If you use the `cursor` option, the LDR job will **not** perform an initial scan, it will stream any changes after the given timestamp. The LDR job will encounter an error if you specify a timestamp with `cursor` that before the configured garbage collection window for that table. {% comment %}to add links to the SQL ref page for create logical rep stream, once published {% endcomment %}
+
+## Schema validation
```
I feel like this section should be in the "setup section" what do you think?

alicia-l2 commented on src/current/v24.3/manage-logical-data-replication.md:
```diff
@@ -0,0 +1,152 @@
+You can run any other CockroachDB [job]({% link {{ page.version.version }}/show-jobs.md %}) on any cluster that is involved in an LDR job. Both source and destination clusters in LDR are active, which means they can both serve production reads and writes as well as run [backups]({% link {{ page.version.version }}/backup-and-restore-overview.md %}), [changefeeds]({% link {{ page.version.version }} ...
+
+{{site.data.alerts.callout_success}}
+In a hot-cold setup with unidirectional LDR streaming from the hot cluster to the cold, you may want to run jobs like changefeeds from the cold cluster to isolate these jobs from the hot cluster receiving application traffic. {% comment %} add link to ldr overview page that will describe this workload isolation topology {% endcomment %}
```
This is a good callout- maybe instead of saying hot-cold, you could just say "In a unidirectional setup you may want to run jobs like X from one cluster, to isolate jobs from the other cluster receiving primary application traffic"

kathancox replied:
Done. I used "principal application traffic", mainly because I was thinking about users searching "primary" or similar in relation to PCR and then ending up here, probably an unfounded concern, but just in case. Happy to use another work though.

alicia-l2 reviewed (commented):
LGTM, with comments!

kathancox commented on src/current/v24.3/manage-logical-data-replication.md:
```diff
@@ -0,0 +1,195 @@
+The schema for the `movr.crdb_replication.dlq_113_public_promo_codes` DLQ:
+
+~~~sql
+CREATE TABLE crdb_replication.dlq_113_public_promo_codes (
```
Shall we add a table to describe each of these columns?

alicia-l2 replied:
Yeah that would be awesome. Ill ask michael or someone from eng to.

kathancox replied:
If you want this in before merging, can we get that today, otherwise let's create an issue for this and the DLQ query.

kathancox commented on src/current/v24.3/manage-logical-data-replication.md:
```diff
@@ -0,0 +1,195 @@
+LDR will not pause when the writes are sent to the DLQ, you must manage the DLQ manually.
+{{site.data.alerts.end}}
+
+To manage the DLQ, you can evaluate entries in the `incoming_row` column and apply the row manually to another table with SQL statements.
```
Any way we could prep a query to help with this before we publish this doc?

alicia-l2 replied:
Yes @msbutler @dt can u help

alicia-l2 commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,213 @@
+_Modes_ determine how LDR replicates the data to the destination cluster. There are two modes:
+
+- `immediate` (default): Attempts to replicate the changed row directly into the destination table, without re-running constraint validations or triggers. It does not support writing into tables with [foreign key]({% link {{ page.version.version }}/foreign-key.md %}) constraints.
+- `validated`: Attempts to apply the write in a similar way to a user-run query, which would re-run all constraint validations or triggers relevant to the destination table(s). It will potentially reject the change if it fails rather than writing it, which will cause the row change to enter the DLQ instead.
```
> It will potentially reject the change if it fails rather than writing it, which will cause the row change to enter the DLQ instead.

I think we should change this sentence to "If the change violates foreign key dependencies, unique constraints, or other constraints, the row will be put in the DLQ instead", because the row will stil be written -- it'll just be written to the DLQ. 

also for this section -- I'm still waiting on eng to provide us with some sample queries the customer can use to get their stuff out of the DLQ. But maybe that would go in the "manage" section

kathancox replied:
Thanks for checking on the queries. And I have changed the language per your rec.

alicia-l2 commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,213 @@
+    CREATE LOGICAL REPLICATION STREAM FROM TABLES ({database.public.table_name},{database.public.table_name},...)  ON 'external://{source_external_connection}' INTO TABLES ({database.public.table_name},{database.public.table_name},...);
+    ~~~
+
+    {% include {{ page.version.version }}/ldr/multiple-tables.md %}
```
I feel like this would be nicer as a note tooltip, like a blue note

alicia-l2 commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,213 @@
+    {% comment  %} TODO: Add output here when example included.
+    To verify that data at a certain point in time is correct on the destination cluster, refer to [Data verification](). 1. After the two jobs are running, you can check the number of rows on each cluster with  {% endcomment %}
+
+## Step 5. Monitor the LDR jobs
```
Should we also call out that metrics export is possible?

kathancox replied:
Added a tip with all the monitoring tools here because the monitoring page won't be published quite a as this page goes out. I will change this when the monitoring page is published soon after.

alicia-l2 commented on src/current/v24.3/manage-logical-data-replication.md:
```diff
@@ -0,0 +1,195 @@
+
+- [Conflict resolution](#conflict-resolution): As changes to a table replicate from the source to the destination cluster, there can be conflicts during some operations that the job will handle with conflict resolution. When LDR is started, the job creates a [_dead letter queue (DLQ)_](#dead-letter-queue-dlq) table with each replicating table. LDR will send any unresolved conflicts to the DLQ, wh ...
+- [Schema changes](#schema-changes): The tables that are part of the LDR job may require schema changes, which need to be coordinated manually. There are some schema changes that are supported while LDR is active.
+- [Jobs](#jobs-and-ldr): Other CockroachDB jobs can operate on clusters running LDR [jobs]({% link {{ page.version.version }}/show-jobs.md %}). You may want to consider where you start and how you manage [backups]({% link {{ page.version.version }}/backup-and-restore-overview.md %}), [changefeeds]({% link {{ page.version.version }}/change-data-capture-overview.md %}), [row-level TTL]({% link {{ p ...
```
"All CockroachDB jobs can still operate on clusters running LDR?"

kathancox replied:
Changed.

alicia-l2 commented on src/current/v24.3/manage-logical-data-replication.md:
```diff
@@ -0,0 +1,195 @@
+Once you have **logical data replication (LDR)** running, you will need to track and manage certain parts of the job:
+
+- [Conflict resolution](#conflict-resolution): As changes to a table replicate from the source to the destination cluster, there can be conflicts during some operations that the job will handle with conflict resolution. When LDR is started, the job creates a [_dead letter queue (DLQ)_](#dead-letter-queue-dlq) table with each replicating table. LDR will send any unresolved conflicts to the DLQ, wh ...
+- [Schema changes](#schema-changes): The tables that are part of the LDR job may require schema changes, which need to be coordinated manually. There are some schema changes that are supported while LDR is active.
```
"while LDR jobs are actively running."?

kathancox replied:
changed

alicia-l2 commented on src/current/v24.3/manage-logical-data-replication.md:
```diff
@@ -0,0 +1,195 @@
+- An `UPDATE` operation replicated to the destination cluster.
+- A cross-cluster write occurs, which means both clusters are writing to the same key. For example, if the LDR stream attempts to apply a row to the destination cluster where the existing row on the destination was not written by the LDR stream.
+
+LDR uses _last write wins (LWW)_ conflict resolution based on the [MVCC timestamp]({% link {{ page.version.version }}/architecture/storage-layer.md %}#mvcc) of the replicating write. LDR will resolve conflicts by inserting the row with the latest MVCC timestamp. When a conflict cannot apply, due to violating a foreign key constraint or schema constraint, it will be retried for up to a minute and  ...
```
"In validated mode, when a conflict cannot apply,"

kathancox replied:
Done

alicia-l2 commented on src/current/v24.3/manage-logical-data-replication.md:
```diff
@@ -0,0 +1,195 @@
+LDR will **not** replicate the allowlist schema changes to the destination table. Therefore, you must perform the schema change carefully on both the source and destination cluster.
+{{site.data.alerts.end}}
+
+### Coordinate schema changes
```
"coordinate other schema changes"

kathancox replied:
done

alicia-l2 commented on src/current/v24.3/manage-logical-data-replication.md:
```diff
@@ -0,0 +1,195 @@
+    ~~~
+
+1. Perform the schema change on cluster A. 
+1. Drop the table from cluster B. 
```
Drop the existing table?

kathancox replied:
Added

alicia-l2 commented on src/current/v24.3/manage-logical-data-replication.md:
```diff
@@ -0,0 +1,195 @@
+    CREATE LOGICAL REPLICATION STREAM FROM TABLE {database.public.table_name} ON 'external://{source_external_connection}' INTO TABLE {database.public.table_name};
+    ~~~
+
+## Jobs and LDR
```
we should call out that Backups function the same as changefeds. like they also contain their application writes, as well as the replicated writes, so u only need to run a backup job on one cluster.

kathancox replied:
I've added a paragraph here on backups, hopefully Michael can check this as he reviews.

alicia-l2 commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,213 @@
+    CREATE EXTERNAL CONNECTION {source} AS 'postgresql://{replication user}:{password}@{node IP}:26257?options=-ccluster%3Dsystem&sslinline=true&sslmode=verify-full&sslrootcert=-----BEGIN+CERTIFICATE-----{encoded certificate}-----END+CERTIFICATE-----%0A`;
+    ~~~
+
+## Step 3. Start LDR
```
ok actually we fixed this restore correctness issue yesterday, so now we can include the cursor bit... So sorry for the whiplast. ARe you able to include the instructions from https://docs.google.com/document/d/1ou6cxZdWiUDC74Zo-dDCTt9IqaDgPf16H3OqdHrwEqo/edit?tab=t.0 for initial bootstrap?

kathancox replied:
I think this might be better placed as an example on the `CREATE LOGICAL REPLICATION STREAM` ref page, as an alternative way to start LDR. BUT, do let me know if you want this to replace what we have in the setup guide right now.

alicia-l2 reviewed (approved):
LGTM -- thank you

kathancox commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,213 @@
+
+1. Prepare the tables on each cluster with the prerequisites for starting LDR.
+1. Set up an [external connection]({% link {{ page.version.version }}/create-external-connection.md %}) on cluster B to hold the connection URI for cluster A.
+1. Start LDR from cluster B with your required modes.
```
Added a small ref to this here.

kathancox commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,213 @@
+1. Set up an [external connection]({% link {{ page.version.version }}/create-external-connection.md %}) on cluster B to hold the connection URI for cluster A.
+1. Start LDR from cluster B with your required modes.
+1. (Optional) Run Steps 2 and 3 again with cluster B as the source and A as the destination, which starts LDR streaming from cluster B to A.
+1. Check the status of the LDR job in the [DB Console]({% link {{ page.version.version }}/ui-overview.md %}).
```
I have added a comment to do this once the monitoring page is merged (after this PR).

kathancox commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,213 @@
+
+You'll need:
+
+- Two separate v24.3 CockroachDB {{ site.data.products.core }} clusters with connectivity between every node in both clusters. The SQL advertise address should be the cluster node advertise address so that the LDR job can plan node-to-node connections between clusters for maximum performance.
```
I think it is good to still call out the version, it is easy for people to not necessarily remember the doc version here. 
I really spelled out the node to node connectivity with another sentence, hope this suffices without being overboard.

kathancox commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,213 @@
+- Two separate v24.3 CockroachDB {{ site.data.products.core }} clusters with connectivity between every node in both clusters. The SQL advertise address should be the cluster node advertise address so that the LDR job can plan node-to-node connections between clusters for maximum performance.
+    - To set up each cluster, you can follow [Deploy CockroachDB on Premises]({% link {{ page.version.version }}/deploy-cockroachdb-on-premises.md %}).
+    - The [Deploy CockroachDB on Premises]({% link {{ page.version.version }}/deploy-cockroachdb-on-premises.md %}) tutorial creates a self-signed certificate for each {{ site.data.products.core }} cluster. To create certificates signed by an external certificate authority, refer to [Create Security Certificates using OpenSSL]({% link {{ page.version.version }}/create-security-certificates-openss ...
+    - Both clusters can be empty.
```
Done, moved stuff round a bit more than you recommended here, but emphasized the table level replication

kathancox commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,213 @@
+
+Before you start LDR, you must ensure that all column names, types, constraints, and unique indexes on the destination table match with the source table.
+
+You cannot use LDR on a table with a schema that contains the following:
```
I would recommend not doing that. It is usually more trouble than it's worth adding "future-looking" language. More that it implies it is something that will change, but we do not know that for sure.

kathancox commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,213 @@
+
+For more details, refer to the LDR [Known limitations]({% link {{ page.version.version }}/set-up-logical-data-replication.md %}#known-limitations).
+
+When you run LDR in `immediate` mode, you cannot replicate a table with [foreign key constraints]({% link {{ page.version.version }}/foreign-key.md %}). In `validated` mode, foreign key constraints **must** match. 
```
Done!

msbutler commented on src/current/v24.3/manage-logical-data-replication.md:
```diff
@@ -0,0 +1,199 @@
+
+## Conflict resolution
+
+Conflicts in LDR are detected when there is either:
```
i think this conflict section would be clearer if we separated it in to a secion on KV level conflicts and sql level conflicts. the UPDATE and cross-cluster write explanations are kv level conficts, as in the LDR is writing to a key that already exists. KV conflicts never end up in the DLQ, as LWW will automatically handle conflict resolution.

In the sql conflict subsection, we can describe how a write may not succeed because it violates a sql constraint, like and FK violation.  sql conflicts may end up in the dlq.

kathancox replied:
I've given this a shot @msbutler please review, and feel free to use the suggestions tool to suggest edits to the markdown, if easier.

msbutler commented on src/current/v24.3/manage-logical-data-replication.md:
```diff
@@ -0,0 +1,199 @@
+
+Allowlist schema change | Exceptions
+-------------------+-----------
+[`CREATE INDEX`]({% link {{ page.version.version }}/create-index.md %}) | <ul><li>Hash-sharded indexes</li><li>Indexes with a computed column</li><li>Partial indexes</li></ul>
```
unique indexes cannot be created mid ldr stream.

kathancox replied:
added

msbutler commented on src/current/v24.3/manage-logical-data-replication.md:
```diff
@@ -0,0 +1,199 @@
+
+### Coordinate other schema changes
+
+To perform any other schema change on the table, use the following approach to redirect application traffic to one cluster. You'll need to drop the existing LDR jobs, perform the schema change, and start new LDR jobs, which will require a full initial scan.
```
once jeff lands and backports some changes, the user will be able to restart the stream with a backup-restore flow.  we could make revisions here once jeff's work lands.

alicia-l2 replied:
I was thinking that we could just document the schema change bit w initial scan to be safer, as opposed to the backup/restore option, if you think b/r is safe too though then sure we can add

kathancox replied:
I haven't changed anything here, because I wasn't sure there was a definite ask, so let me know if I do need to do anything.

msbutler replied:
i no longer think we should doc b/r here.

msbutler commented on src/current/v24.3/manage-logical-data-replication.md:
```diff
@@ -0,0 +1,199 @@
+
+## Jobs and LDR
+
+You can run any other CockroachDB [job]({% link {{ page.version.version }}/show-jobs.md %}) on any cluster that is involved in an LDR job. Both source and destination clusters in LDR are active, which means they can both serve production reads and writes as well as run [backups]({% link {{ page.version.version }}/backup-and-restore-overview.md %}), [changefeeds]({% link {{ page.version.version }} ...
```
"can run any other job" is not quite right, for example, you cannot run import on an LDR job right now. 

i would only discuss backup and changefeed here.

kathancox replied:
Specified cdc/ backup here

msbutler commented on src/current/v24.3/manage-logical-data-replication.md:
```diff
@@ -0,0 +1,199 @@
+You can run any other CockroachDB [job]({% link {{ page.version.version }}/show-jobs.md %}) on any cluster that is involved in an LDR job. Both source and destination clusters in LDR are active, which means they can both serve production reads and writes as well as run [backups]({% link {{ page.version.version }}/backup-and-restore-overview.md %}), [changefeeds]({% link {{ page.version.version }} ...
+
+{{site.data.alerts.callout_success}}
+In a unidirectional setup, you may want to run jobs like [changefeeds]({% link {{ page.version.version }}/change-data-capture-overview.md %}) from one cluster to isolate these jobs from the cluster receiving the principal application traffic. {% comment %} add link to ldr overview page that will describe this workload isolation topology {% endcomment %}
```
you could want changefeeds on one cluster in a bidi hot cold setup too? I would reframe this as, "if one cluster in your ldr setup has a main traffic, you could run backups/changefeeds on the other cluster"

kathancox replied:
Removed "unidirectional" here

msbutler commented on src/current/v24.3/manage-logical-data-replication.md:
```diff
@@ -0,0 +1,199 @@
+
+[Changefeeds]({% link {{ page.version.version }}/change-data-capture-overview.md %}) will emit [messages]({% link {{ page.version.version }}/changefeed-messages.md %}) for the writes that occur to the watched table. If the watched table is also the destination to which LDR is streaming, the changefeed will additionally emit messages for the writes from the LDR job. For example:
+
+1. You create a changefeed watching the `test_table` on cluster A.
```
why do we need this extra example? i think the paragraph above explains this well.

alicia-l2 replied:
Eh I feel like we could keep this, sometimes ppl still get confused

kathancox replied:
I think J4 seemed to be happy with this additional example, so it's probably worth keeping it in

msbutler reviewed (commented):
Thanks for all your work here @kathancox !! I realize it is very hard to write docs for LDR because of how large of a surface area it has and how much last minute dev work we're doing. I really appreciate your work! 

Left a lot of comments, but i think this is nearly there.

alicia-l2 commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,216 @@
+- LDR replicates at the table level, which means clusters can contain other tables that are not part of the LDR job. If both clusters are empty, create the tables that you need to replicate with **identical** schema definitions (excluding indexes) on both clusters. If one cluster already has an existing table that you'll replicate, ensure the other cluster's table definition matches. For more det ...
+
+{{site.data.alerts.callout_info}}
+If you need to run LDR through a load balancer, use the load balancer IP address as the SQL advertise address on each cluster. It is important to note that using a load balancer with LDR can impair performance.
```
Yea

kathancox replied:
Commented out for now

kathancox commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,216 @@
+
+When you run LDR in `immediate` mode, you cannot replicate a table with [foreign key constraints]({% link {{ page.version.version }}/foreign-key.md %}). In `validated` mode, foreign key constraints **must** match. All constraints are enforced at the time of SQL/application write.
+
+## Step 1. Prepare the cluster
```
There is the tutorial overview section that runs through this and mentions the optional step. I will link to additional pages on the overviews of setups as things merge. I also think the page's introduction kind of covers this too.

kathancox commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,216 @@
+
+    You can change the default `mode` using the `WITH mode = validated` syntax.
+
+    If you would like to add multiple tables to the LDR job:
```
Done

kathancox commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,216 @@
+If you would like to ignore TTL deletes in LDR, you can use the `discard = ttl-deletes` option in the `CREATE LOGICAL REPLICATION STREAM` statement.
+{{site.data.alerts.end}}
+
+1. From the **destination** cluster, start LDR. Use the fully qualified table name and ensure the source and destination table names are identical:
```
OK I removed this, and added in `source` /` `destination` to the variables for clarity here.

kathancox commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,216 @@
+- `validated`: Attempts to apply the write in a similar way to a user-run query, which would re-run all constraint validations or triggers relevant to the destination table(s). If the change violates foreign key dependencies, unique constraints, or other constraints, the row will be put in the DLQ instead.
+
+{{site.data.alerts.callout_info}}
+If you would like to ignore TTL deletes in LDR, you can use the `discard = ttl-deletes` option in the `CREATE LOGICAL REPLICATION STREAM` statement.
```
I have left this note in, but I'm ok to remove this too, if we don't want to "advertise" this here.

msbutler replied:
@alicia-l2 to decide if we should leave this not in. i don't want it :D

kathancox commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,216 @@
+
+_Modes_ determine how LDR replicates the data to the destination cluster. There are two modes:
+
+- `immediate` (default): Attempts to replicate the changed row directly into the destination table, without re-running constraint validations or triggers. It does not support writing into tables with [foreign key]({% link {{ page.version.version }}/foreign-key.md %}) constraints.
```
Removed

kathancox commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,216 @@
+
+    If you need to change the password later, refer to [`ALTER USER`]({% link {{ page.version.version }}/alter-user.md %}). {% comment  %}Add link to https://www.cockroachlabs.com/docs/stable/ui-overview#db-console-security-considerations for secure clusters{% endcomment %}
+
+1. (Optional) If you would like to ignore row-level TTL deletes in LDR, set the `ttl_disable_changefeed_replication` storage parameter on the table. On the **source** cluster, alter the table to set the table storage parameter:
```
Removed from here.

kathancox commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,216 @@
+    cockroach sql --url "postgresql://root@{node IP or hostname}:26257?sslmode=verify-full" --certs-dir=certs
+    ~~~
+
+1. On **both** clusters, enable the `kv.rangefeed.enabled` cluster setting:
```
Added clarification here.

kathancox commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,216 @@
+- Columns with [user-defined types]({% link {{ page.version.version }}/create-type.md %})
+- [Column families]({% link {{ page.version.version }}/column-families.md %})
+- [Partial indexes]({% link {{ page.version.version }}/partial-indexes.md %})
+- Indexes with a [computed column]({% link {{ page.version.version }}/computed-columns.md %})
```
Done

kathancox commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,216 @@
+
+- Columns with [user-defined types]({% link {{ page.version.version }}/create-type.md %})
+- [Column families]({% link {{ page.version.version }}/column-families.md %})
+- [Partial indexes]({% link {{ page.version.version }}/partial-indexes.md %})
```
Added

kathancox commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,216 @@
+
+You cannot use LDR on a table with a schema that contains the following:
+
+- Columns with [user-defined types]({% link {{ page.version.version }}/create-type.md %})
```
Removed

kathancox commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,216 @@
+
+You'll need:
+
+- Two separate v24.3 CockroachDB {{ site.data.products.core }} clusters with connectivity between every node in both clusters. That is, all nodes in cluster A must be able to contact each node in cluster B and vice versa. The SQL advertise address should be the cluster node advertise address so that the LDR job can plan node-to-node connections between clusters for maximum performance.
```
done

kathancox commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,216 @@
+Logical data replication is only supported in CockroachDB {{ site.data.products.core }} clusters.
+{{site.data.alerts.end}}
+
+In this tutorial, you will set up **logical data replication (LDR)** streaming data from a source table to a destination table between two CockroachDB clusters. Both clusters are active and can serve traffic. You can apply the outlined steps to create _unidirectional_ LDR from a source table to a destination table (cluster A to cluster B) in one LDR job. Optionally, you can also create _bidirecti ...
```
Yes, trying to keep some consistency between job, stream, plain LDR, but there is a mix in the syntax, so there is a mix here too. I'll discuss this a bit with my docs reviewer too.

msbutler replied:
sg. happy with whatever y'all decide.

msbutler commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -67,7 +66,7 @@ When you run LDR in `immediate` mode, you cannot replicate a table with [foreign
     ~~~
 
-1. On **both** clusters, enable the `kv.rangefeed.enabled` cluster setting:
+1. If you are setting up bidirectional LDR, on **both** clusters, enable the `kv.rangefeed.enabled` cluster setting. If you are setting up unidirectional LDR, enable this setting on the **source** cluster:
```
given that subsequent steps are written for a unidirectional setup, and that we ask the user in a bidi setup to rinse and repeat below, i think it might be clearer to just say the source needs to enable rangefeeds here.

msbutler reviewed (approved):
Looks good! Just a few nits and one question for alicia

rmloveland commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,206 @@
+
+You'll need:
+
+- Two separate v24.3 CockroachDB {{ site.data.products.core }} clusters with connectivity between every node in both clusters. That is, all nodes in cluster A must be able to contact each node in cluster B and vice versa. The SQL advertised address should be the cluster node advertise address so that the LDR job can plan node-to-node connections between clusters for maximum performance.
```
suggest replacing "v24.3" with `{{ page.version.version }}` variable so it doesn't need to be touched again later and/or for backports, new version publishing etc.

rmloveland commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,206 @@
+- Two separate v24.3 CockroachDB {{ site.data.products.core }} clusters with connectivity between every node in both clusters. That is, all nodes in cluster A must be able to contact each node in cluster B and vice versa. The SQL advertised address should be the cluster node advertise address so that the LDR job can plan node-to-node connections between clusters for maximum performance.
+    - To set up each cluster, you can follow [Deploy CockroachDB on Premises]({% link {{ page.version.version }}/deploy-cockroachdb-on-premises.md %}).
+    - The [Deploy CockroachDB on Premises]({% link {{ page.version.version }}/deploy-cockroachdb-on-premises.md %}) tutorial creates a self-signed certificate for each {{ site.data.products.core }} cluster. To create certificates signed by an external certificate authority, refer to [Create Security Certificates using OpenSSL]({% link {{ page.version.version }}/create-security-certificates-openss ...
+    - All nodes in each cluster will need access to the Certificate Authority for the other cluster. Refer to [Connect from the destination to the source](#step-2-connect-from-the-destination-to-the-source).
```
suggest making link name "Step 2. connect from dest to src" - i was confused if this was linking to a step or another page

kathancox replied:
Done!

rmloveland commented on src/current/v24.3/set-up-logical-data-replication.md:
```diff
@@ -0,0 +1,206 @@
+_Modes_ determine how LDR replicates the data to the destination cluster. There are two modes:
+
+- `immediate` (default): Attempts to replicate the changed row directly into the destination table, without re-running constraint validations. It does not support writing into tables with [foreign key]({% link {{ page.version.version }}/foreign-key.md %}) constraints.
+- `validated`: Attempts to apply the write in a similar way to a user-run query, which would re-run all constraint validations or triggers relevant to the destination table(s). If the change violates foreign key dependencies, unique constraints, or other constraints, the row will be put in the DLQ instead.
```
suggest spelling out first use of DLQ (or linking to something if we have)

kathancox replied:
Ooops, thanks for catching that!

rmloveland commented on src/current/v24.3/manage-logical-data-replication.md:
```diff
@@ -0,0 +1,207 @@
+
+- [Conflict resolution](#conflict-resolution): As changes to a table replicate from the source to the destination cluster, there can be conflicts during some operations that the job will handle with conflict resolution. When LDR is started, the job creates a [_dead letter queue (DLQ)_](#dead-letter-queue-dlq) table with each replicating table. LDR will send any unresolved conflicts to the DLQ, wh ...
+- [Schema changes](#schema-changes): The tables that are part of the LDR job may require schema changes, which need to be coordinated manually. There are some schema changes that are supported while LDR jobs are actively running.
+- [Jobs](#jobs-and-ldr): All CockroachDB jobs can operate on clusters running LDR [jobs]({% link {{ page.version.version }}/show-jobs.md %}). You may want to consider where you start and how you manage [backups]({% link {{ page.version.version }}/backup-and-restore-overview.md %}), [changefeeds]({% link {{ page.version.version }}/change-data-capture-overview.md %}), [row-level TTL]({% link {{ pag ...
```
"CockroachDB jobs" seems nonsensical since they all are? do you mean non-LDR jobs?

kathancox replied:
Oh yeah! Great point. I have changed to changefeeds and backups as those are the jobs we want to discuss here.

rmloveland commented on src/current/v24.3/manage-logical-data-replication.md:
```diff
@@ -0,0 +1,207 @@
+
+## Jobs and LDR
+
+You can run changefeed and backup CockroachDB [jobs]({% link {{ page.version.version }}/show-jobs.md %}) on any cluster that is involved in an LDR job. Both source and destination clusters in LDR are active, which means they can both serve production reads and writes as well as run [backups]({% link {{ page.version.version }}/backup-and-restore-overview.md %}), [changefeeds]({% link {{ page.versi ...
```
same as other comment re: do you mean "non-LDR jobs"?

rmloveland reviewed (approved):
LGTM! some non-blocking comments, take or leave

kathancox commented:
TFTRs! Merging this, more updates to LDR docs to come shortly.
